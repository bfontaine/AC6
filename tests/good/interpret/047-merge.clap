** Generate a list of all number from 1 to n.
def generate (n) =
  if n = 0 then Nil else Cons [{ Hd <- n, Tl <- generate (n - 1) }]

** [insert i l] where [l] is sorted, returns a list where
** [i] is at the right position.
def insert (i) (l) =
  l . case {
    Nil => 
    Cons [{ Hd <- i, Tl <- Nil }]
  | Cons [{ Hd -> x, Tl -> xs }] =>
    if i < x then 
      Cons [{ Hd <- i, Tl <- l }] 
    else 
      Cons [{ Hd <- x, Tl <- insert i xs }]
}     

** Tail recursive version of insertion sort.
def sort_aux (accu) (l) = 
 l . case {
    Nil => accu
  | Cons [{ Hd -> x, Tl -> xs }] =>
    sort_aux (insert x accu) xs
 }

def sort (l) = sort_aux (Nil) l

** Bench
val main = 
  val l = generate 10000 in 
  val ls = sort l in
  Unit

