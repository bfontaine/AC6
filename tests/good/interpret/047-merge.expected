:- generate = <code>
:- insert = <code>
:- sort_aux = <code>
:- sort = <code>
:- main = Unit
